tests/mbt/model/ccv.qnt

```diff
@@ -654,6 +840,168 @@ module ccv {
+        val consumerState = currentState.consumerStates.get(consumer)
+        if (not(isRunningConsumer(consumer, currentState.providerState))) {
+            Err("Consumer is not currently a consumer - must have 'running' status!")
+        } else if (downtime and consumerState.outstandingDowntime.contains(consumerNode)) {
+            Err("Consumer already has outstanding downtime for validator!")
+        } else {
+            val vscPacket = consumerState.receivedVscPackets.select(packet => packet.id == vscId)[0]
+            val packet = Slash({
+                validator: consumerNode,
+                id: vscId,
+                downtime: downtime,
+                sendingTime: consumerState.chainState.runningTimestamp,
+                timeoutTime: consumerState.chainState.runningTimestamp + CcvTimeout.get(consumer),
+                valPower: vscPacket.validatorSet.get(consumerNode)
+
+            })
+            val newConsumerState = 
+                commitPacketOnConsumer(consumerState, packet)
+                // add the consumer address to the list of outstanding slashes
+                .with("outstandingDowntime", consumerState.outstandingDowntime.union(Set(consumerNode)))
+            Ok(currentState.with("consumerStates", currentState.consumerStates.set(consumer, newConsumerState)))
+        }
+    }
+
+    // receives a given packet on the provider. The arguments are the consumer chain that sent the packet, and the packet itself.
+    pure def recvPacketOnProvider(currentState: ProtocolState, sender: Chain, packet: Packet): Result = {
+        match packet {
+            | VscMatured(p) => recvVscMaturedPacketOnProvider(currentState, sender, p)
+            | Slash(p) => recvSlashPacketOnProvider(currentState, sender, p)
+            | _ => Err("Do not know how to handle this packet type")
+        }
+    }
+
+    pure def recvSlashPacketOnProvider(currentState: ProtocolState, sender: Chain, packet: SlashPacket): Result = {
+        // determine the jail duration for the validator:
+        // if the packet is a downtime slash, the jail duration is DowntimeJailTime
+        // otherwise, the validator is jailed until the DoubleSignJailTime
+        val jailEndTime = if (packet.downtime) {
+            packet.sendingTime + DowntimeJailTime
+        } else {
+            packet.sendingTime + DoubleSignJailEndTime
+        }
+        // map the validator to its address on the provider
+        // (in general, the validator is a consumer address and could be an assigned one)
+        val providerVal = currentState.providerState.consumerAddrToValidator
+            .getOrElse(sender, Map())
+            // if there is no entry for that validator, there is no key assignment
+            .getOrElse(packet.validator, packet.validator)
+
+        val slashFactor = if (packet.downtime) {
+            DowntimeSlashPercentage
+        } else {
+            DoubleSignSlashPercentage
+        }
+        
+        val newProviderState = 
```

Comment:
Shouldn't we check that the validator isn't already jailed before?

Reply:
Great point - I didn't catch this because I don't compare the jail times, will add this check in the driver and also adjust this here